File: lib/timer-logic.js
// Pure functions for the pomodoro state machine and weight system.
// Kept dependency-free so they run in the service worker, content scripts,
// the popup, and the node test runner.

(function (root) {
  const DEFAULT_WORK_MINUTES = 25;
  const BREAK_MINUTES = 5;
  const MIN_WORK_MINUTES = 1;
  const MAX_WORK_MINUTES = 120;

  // Weight stages as a function of continuous work hours.
  // A "stage" determines how much of the screen the cat occupies.
  // Stage maps to a viewport-relative size factor (0.20 = 20% of min(vw,vh)).
  function weightStageForHours(hours) {
    if (hours < 1) return { stage: 0, sizeFactor: 0.22, label: "smug" };
    if (hours < 2) return { stage: 1, sizeFactor: 0.32, label: "chunky" };
    if (hours < 3) return { stage: 2, sizeFactor: 0.45, label: "rotund" };
    if (hours < 5) return { stage: 3, sizeFactor: 0.60, label: "obscene" };
    return { stage: 4, sizeFactor: 0.85, label: "eclipse" };
  }

  // Stretch factor during the break window. At 5:00 the cat is compact;
  // it stretches out as the break ticks down to 4:47 (the "getting comfortable" arc).
  // After 4:47 it stays stretched for the rest of the break.
  function stretchFactor(secondsRemaining, breakSeconds) {
    const totalStretchWindow = 13; // 5:00 -> 4:47
    const elapsed = breakSeconds - secondsRemaining;
    if (elapsed <= 0) return 1.0;
    if (elapsed >= totalStretchWindow) return 1.6;
    return 1.0 + (elapsed / totalStretchWindow) * 0.6;
  }

  function clampWorkMinutes(n) {
    const v = Math.floor(Number(n));
    if (!Number.isFinite(v)) return DEFAULT_WORK_MINUTES;
    return Math.min(MAX_WORK_MINUTES, Math.max(MIN_WORK_MINUTES, v));
  }

  function formatMMSS(seconds) {
    const s = Math.max(0, Math.floor(seconds));
    const m = Math.floor(s / 60);
    const r = s % 60;
    return `${m}:${r.toString().padStart(2, "0")}`;
  }

  // Compute remaining seconds in the current phase given a timestamp.
  // state: { phase: 'work'|'break'|'idle', phaseEndsAt: epochMs }
  function remainingSeconds(state, nowMs) {
    if (!state || state.phase === "idle") return 0;
    return Math.max(0, Math.ceil((state.phaseEndsAt - nowMs) / 1000));
  }

  // Continuous work hours used to calculate weight.
  // workStreakStartedAt resets only when a break is actually completed.
  function continuousWorkHours(workStreakStartedAt, nowMs) {
    if (workStreakStartedAt === null || workStreakStartedAt === undefined) return 0;
    return Math.max(0, (nowMs - workStreakStartedAt) / (1000 * 60 * 60));
  }

  // Determine the next state when the current phase elapses.
  function advancePhase(state, settings, nowMs) {
    const workMs = clampWorkMinutes(settings.workMinutes) * 60 * 1000;
    const breakMs = BREAK_MINUTES * 60 * 1000;
    if (state.phase === "work") {
      const streak =
        state.workStreakStartedAt === null || state.workStreakStartedAt === undefined
          ? nowMs
          : state.workStreakStartedAt;
      return {
        phase: "break",
        phaseEndsAt: nowMs + breakMs,
        workStreakStartedAt: streak,
        breakStartedAt: nowMs,
      };
    }
    if (state.phase === "break") {
      // Break completed -> reset streak, start a fresh work cycle.
      return {
        phase: "work",
        phaseEndsAt: nowMs + workMs,
        workStreakStartedAt: nowMs,
        breakStartedAt: null,
      };
    }
    // idle -> work
    return {
      phase: "work",
      phaseEndsAt: nowMs + workMs,
      workStreakStartedAt: nowMs,
      breakStartedAt: null,
    };
  }

  const api = {
    DEFAULT_WORK_MINUTES,
    BREAK_MINUTES,
    MIN_WORK_MINUTES,
    MAX_WORK_MINUTES,
    weightStageForHours,
    stretchFactor,
    clampWorkMinutes,
    formatMMSS,
    remainingSeconds,
    continuousWorkHours,
    advancePhase,
  };

  if (typeof module !== "undefined" && module.exports) {
    module.exports = api;
  } else {
    root.CatTimer = api;
  }
})(typeof self !== "undefined" ? self : this);

File: android-app/app/src/main/kotlin/com/fatorangecat/core/TimerLogic.kt
package com.fatorangecat.core

import kotlin.math.ceil
import kotlin.math.floor
import kotlin.math.max
import kotlin.math.min

// Pure functions for the pomodoro state machine and weight system.
// Mirrors chrome-extension/lib/timer-logic.js line-for-line. No Android
// dependencies — runnable on plain JVM JUnit.

object TimerLogic {

    const val DEFAULT_WORK_MINUTES: Int = 25
    const val BREAK_MINUTES: Int = 5
    const val MIN_WORK_MINUTES: Int = 1
    const val MAX_WORK_MINUTES: Int = 120
    const val WARNING_SECONDS: Int = 30

    enum class Phase { IDLE, WORK, BREAK }

    data class TimerState(
        val phase: Phase,
        val phaseEndsAt: Long = 0L,
        val workStreakStartedAt: Long? = null,
        val breakStartedAt: Long? = null,
    )

    data class Settings(val workMinutes: Any?)

    data class WeightStage(val stage: Int, val sizeFactor: Double, val label: String)

    fun weightStageForHours(hours: Double): WeightStage {
        if (hours < 1.0) return WeightStage(0, 0.22, "smug")
        if (hours < 2.0) return WeightStage(1, 0.32, "chunky")
        if (hours < 3.0) return WeightStage(2, 0.45, "rotund")
        if (hours < 5.0) return WeightStage(3, 0.60, "obscene")
        return WeightStage(4, 0.85, "eclipse")
    }

    fun stretchFactor(secondsRemaining: Int, breakSeconds: Int): Double {
        val totalStretchWindow = 13
        val elapsed = breakSeconds - secondsRemaining
        if (elapsed <= 0) return 1.0
        if (elapsed >= totalStretchWindow) return 1.6
        return 1.0 + (elapsed.toDouble() / totalStretchWindow) * 0.6
    }

    fun clampWorkMinutes(input: Any?): Int {
        val asDouble: Double = when (input) {
            is Number -> input.toDouble()
            is String -> input.toDoubleOrNull() ?: return DEFAULT_WORK_MINUTES
            else -> return DEFAULT_WORK_MINUTES
        }
        if (asDouble.isNaN() || asDouble.isInfinite()) return DEFAULT_WORK_MINUTES
        val floored = floor(asDouble).toInt()
        return min(MAX_WORK_MINUTES, max(MIN_WORK_MINUTES, floored))
    }

    fun formatMMSS(seconds: Number): String {
        val s = max(0, floor(seconds.toDouble()).toInt())
        val m = s / 60
        val r = s % 60
        return "$m:${r.toString().padStart(2, '0')}"
    }

    fun remainingSeconds(state: TimerState, nowMs: Long): Int {
        if (state.phase == Phase.IDLE) return 0
        val delta = state.phaseEndsAt - nowMs
        return max(0, ceil(delta / 1000.0).toInt())
    }

    fun shouldWarn(state: TimerState, nowMs: Long): Boolean {
        if (state.phase != Phase.WORK) return false
        val r = remainingSeconds(state, nowMs)
        return r in 1..WARNING_SECONDS
    }

    fun continuousWorkHours(workStreakStartedAt: Long?, nowMs: Long): Double {
        if (workStreakStartedAt == null) return 0.0
        return max(0.0, (nowMs - workStreakStartedAt) / (1000.0 * 60.0 * 60.0))
    }

    fun advancePhase(state: TimerState, settings: Settings, nowMs: Long): TimerState {
        val workMs = clampWorkMinutes(settings.workMinutes) * 60L * 1000L
        val breakMs = BREAK_MINUTES * 60L * 1000L
        return when (state.phase) {
            Phase.WORK -> {
                val streak = state.workStreakStartedAt ?: nowMs
                TimerState(
                    phase = Phase.BREAK,
                    phaseEndsAt = nowMs + breakMs,
                    workStreakStartedAt = streak,
                    breakStartedAt = nowMs,
                )
            }
            Phase.BREAK -> TimerState(
                phase = Phase.WORK,
                phaseEndsAt = nowMs + workMs,
                workStreakStartedAt = nowMs,
                breakStartedAt = null,
            )
            Phase.IDLE -> TimerState(
                phase = Phase.WORK,
                phaseEndsAt = nowMs + workMs,
                workStreakStartedAt = nowMs,
                breakStartedAt = null,
            )
        }
    }
}
